hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientClusterServiceImpl.java

```diff
@@ -174,12 +177,15 @@ private String addMembershipListenerWithoutInit(MembershipListener listener) {
     }
 
     private void initMembershipListener() {
-        for (MembershipListener membershipListener : listeners.values()) {
-            if (membershipListener instanceof InitialMembershipListener) {
-                // TODO: needs sync with membership events...
-                Cluster cluster = client.getCluster();
-                InitialMembershipEvent event = new InitialMembershipEvent(cluster, cluster.getMembers());
-                ((InitialMembershipListener) membershipListener).init(event);
+        synchronized (initialMembershipListenerMutex) {
+            for (MembershipListener membershipListener : listeners.values()) {
+                if (membershipListener instanceof InitialMembershipListener) {
+                    Cluster cluster = client.getCluster();
+                    Collection<Member> memberCollection = members.get().values();
+                    LinkedHashSet<Member> members = new LinkedHashSet<Member>(memberCollection);
+                    InitialMembershipEvent event = new InitialMembershipEvent(cluster, members);
+                    ((InitialMembershipListener) membershipListener).init(event);
```

Comment:
seems partly duplicate code as addMembershipListener, could be refactored into a common method